File: compiler/src/span.ts
export interface Span {
    path: string;
    start: Location;
    end: Location;
    source: string;
}

export interface Location {
    line: number;
    column: number;
    offset: number;
}

export const nullSpan = (path: string): Span => ({
    path,
    source: "",
    start: { line: 1, column: 1, offset: 0 },
    end: { line: 1, column: 1, offset: 0 },
});

export const compareSpans = (a: Span, b: Span): number => {
    if (a.path !== b.path) {
        return a.path.localeCompare(b.path);
    }

    if (a.start.offset !== b.start.offset) {
        return a.start.offset - b.start.offset;
    }

    return a.end.offset - b.end.offset;
};

export const joinSpans = (left: Span, right: Span, source: string): Span => ({
    path: left.path,
    start: left.start,
    end: right.end,
    source: source.slice(left.start.offset, right.end.offset),
});

export const displaySpan = (span: Span): string =>
    `${span.path}:${span.start.line}:${span.start.column}`;

File: compiler/database/span.go
package database

import (
	"fmt"
	"strings"
)

type Span struct {
	Path   string
	Start  Location
	End    Location
	Source string
}

type Location struct {
	Line   int
	Column int
	Index  int
}

func JoinSpans(left Span, right Span, source string) Span {
	return Span{
		Path:   left.Path,
		Start:  left.Start,
		End:    right.End,
		Source: source[left.Start.Index:right.End.Index],
	}
}

func CompareSpans(left Span, right Span) int {
	if left.Path != right.Path {
		return strings.Compare(left.Path, right.Path)
	}

	if left.Start.Index != right.Start.Index {
		return left.Start.Index - right.Start.Index
	}

	if left.End.Index != right.End.Index {
		return left.End.Index - right.End.Index
	}

	return 0
}

func SpansAreEqual(left Span, right Span) bool {
	return CompareSpans(left, right) == 0
}

func HaveEqualSpans(left Node, right Node) bool {
	return SpansAreEqual(GetSpanFact(left), GetSpanFact(right))
}

func (span Span) String() string {
	return fmt.Sprintf("%s:%d:%d", span.Path, span.Start.Line, span.Start.Column)
}

func NullSpan() Span {
	return Span{
		Path:   "",
		Start:  NullLocation(),
		End:    NullLocation(),
		Source: "",
	}
}

func NullLocation() Location {
	return Location{
		Line:   1,
		Column: 1,
		Index:  0,
	}
}
